src/boat_simulator/boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,260 @@
+    ) -> Scalar:  # angle passed in as radians
+        """Computes the error between desired and current heading
+        implementation taken from: https://stackoverflow.com/a/2007279
+        Angles are bound with the convention (-180, 180]
+
+        Args:
+            `desired` (Scalar): New desired heading in degrees
+            `current` (Scalar): New current heading in degrees
+
+        Returns:
+            Scalar: The error between the given headings in radians.
+        """
+        current_bound, desired_bound = map(bound_to_180, (current, desired))
+        desired_rad, current_rad = np.deg2rad(desired_bound), np.deg2rad(current_bound)
+
+        error = atan2(sin(desired_rad - current_rad), cos(desired_rad - current_rad))
+
+        return error
+
+    def compute_error_angle(self) -> Scalar:
+        """Computes the feedback angle from the given error (in radians)
+        Calculations done using Raye's implementation
+            https://github.com/UBCSailbot/raye-boat-controller/blob/master/python/tack_controller.py
+
+        Returns:
+            Scalar: Difference between current and target rudder angle in degrees.
+        """
+        error = self.compute_error(self.current_heading, self.desired_heading)
+
+        if abs(error) > (pi / 4):  # limited between (-45, 45)
+            raise ValueError("Heading_error must be between -pi/4 and pi/4")
+
+        rudder_change = (self.kp * error) / (1 + (self.cp * error))
+
+        if abs(rudder_change) > (pi / 2):  # greatest change is 90 degrees
+            raise ValueError("Rudder_change cannot be greater than pi / 2")
+
+        rudder_change_deg = np.rad2deg(rudder_change)
+        self.running_error = rudder_change_deg
+
+        return rudder_change_deg
+
+    def compute_setpoint(self) -> Scalar:
+        """Calculates the rudder setpoint angle. Only called at the very beginning once
+        to determine the setpoint angle.
+
+        Returns:
+            The rudder setpoint angle in degrees
+        """
+        return self.current_control_ang + self.compute_error_angle()
+
+    def reset_desired_heading(self, target_angle: Scalar) -> None:
+        """Resets a new desired heading angle. Must recompute all errors and
+        feedback angles again for new iteration loop.
+
+        Args:
+            `target_angle` (Scalar): New desired heading
+
+        """
+        self.desired_heading = target_angle
+
+    def update_state(self) -> bool:
+        """Updates the rudder angle iteratively towards the target rudder setpoint based
+        on the rudder_speed and time step parameters
+
+        Returns:
+            False if target angle was not reached, else True if desired heading has been reached
+        """
+
+        if self.running_error > 0:
+            change = self.rudder_speed * self.time_step
+        elif self.running_error < 0:
+            change = -self.rudder_speed * self.time_step
+        else:
+            return True
```

Comment:
I would be careful trying to compare this way because you assume that you can get exactly zero. But you might get numerical issues with floating point numbers, so try and compare within a tolerance rather than exact equality.

You can use `math.copysign` to get the sign of a number rather than using this `if-else` statement as well.

First see if you are close to zero, otherwise do the actuation.